openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -65,6 +65,57 @@ use x86defs::cpuid::CpuidFunction;
 use zerocopy::FromZeros;
 use zerocopy::IntoBytes;
 
+/// Error type for proxy interrupt redirection failures.
+#[derive(Debug, thiserror::Error)]
+enum ProxyInterruptRedirectionError {
+    /// Multicast is not supported for proxy interrupt redirection.
+    #[error("multicast not supported")]
+    MulticastNotSupported,
+    /// Processor set operation failed.
+    #[error("processor set operation failed")]
+    ProcessorSetError,
+    /// Failed to map the redirected device interrupt in VTL2 kernel.
+    #[error("failed to map redirected device interrupt in VTL2 kernel")]
+    MapInterruptFailed,
+    /// HvCallRetargetDeviceInterrupt hypercall failed in hypervisor.
+    #[error("HvCallRetargetDeviceInterrupt with proxy redirect failed in hypervisor")]
+    RetargetDeviceInterruptFailed,
+}
+
+/// Manages redirected interrupt vector mapping in VTL2 for proxy interrupt redirection.
+struct RedirectedVectorMapping<'a> {
+    hcl: &'a hcl::ioctl::Hcl,
+    apic_id: u32,
+    redirected_vector: u32,
+}
+
+impl<'a> RedirectedVectorMapping<'a> {
+    /// Creates a new mapping in VTL2 kernel for proxy interrupt redirection. This will be automatically
+    /// unmapped when the guard is dropped unless explicitly disarmed.
+    fn new(hcl: &'a hcl::ioctl::Hcl, vector: u32, apic_id: u32) -> Option<Self> {
+        let redirected_vector = hcl.map_redirected_device_interrupt(vector, apic_id, true)?;
+        Some(Self {
+            hcl,
+            apic_id,
+            redirected_vector,
+        })
+    }
+
+    /// Returns the redirected vector value returned by the VTL2 kernel.
+    fn redirected_vector(&self) -> u32 {
+        self.redirected_vector
+    }
+}
+
+impl Drop for RedirectedVectorMapping<'_> {
+    /// Drop guard to unmap the interrupt vector in VTL2 kernel.
+    fn drop(&mut self) {
+        self.hcl
+            .map_redirected_device_interrupt(self.redirected_vector, self.apic_id, false)
+            .expect("Failed to unmap VTL2 vector for proxy device interrupt");
```

Comment:
nit: lowercase for "expect" as well. I think in this case it may be useful to know what redirected vector and apic_id failed, in which case I would probably switch this from an expect to an explicit match call and a `panic!`